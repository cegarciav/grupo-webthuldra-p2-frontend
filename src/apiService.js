import axios from 'axios';

const apiOrigin = process.env.REACT_APP_API_ORIGIN;

const apiRequester = axios.create({
  baseURL: apiOrigin,
});

const errorGenerator = (e) => {
  let errorResponse = null;
  if (e.response) {
    errorResponse = {
      errors: e.response.data.errors,
      type: 'response',
      statusCode: e.response.status,
    };
  } else {
    errorResponse = {
      type: 'unexpected',
      message: 'Algo salió mal. Por favor, revisa tu conexión o intenta ingresar más tarde',
    };
  }
  return errorResponse;
};

async function apiGet(path, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.get(
      url, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPost(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.post(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiPatch(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.patch(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

async function apiDelete(path, body, params) {
  try {
    if (!apiOrigin) throw Error('La conexión con el servidor no ha podido ser establecida');
    const url = new URL(`${apiOrigin}${path}`);
    const token = localStorage.getItem('apiToken');
    let headers = {};
    if (token) headers = { Authorization: `bearer ${token}` };
    if (params) {
      Object.keys(params)
        .forEach((key) => url.searchParams.append(key, params[key]));
    }
    const response = await apiRequester.delete(
      url,
      body, {
        headers,
      },
    );
    return {
      data: response.data,
      statusCode: response.status,
    };
  } catch (e) {
    return errorGenerator(e);
  }
}

export {
  apiGet,
  apiPost,
  apiPatch,
  apiDelete,
};
