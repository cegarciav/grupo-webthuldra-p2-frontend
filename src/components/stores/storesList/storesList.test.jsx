import React from 'react';
import PropTypes from 'prop-types';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import MainRouter from '../../../routers/index';
import { apiGet } from '../../../apiService';

function TestRouter({ path }) {
  return (
    <MemoryRouter initialEntries={[path]}>
      <MainRouter />
    </MemoryRouter>
  );
}

TestRouter.propTypes = {
  path: PropTypes.string.isRequired,
};

jest.mock('../../../apiService');

const storesResponse = [
  {
    id: '6f4e5985-2ff8-4118-bbaf-e6004d757c13',
    address: 'Address 1',
    name: 'Store name 1',
    description: 'Description 1',
    ownerId: '7d13c742-cd2d-45b7-9866-20dbb0eaa2e9',
    picture: null,
    createdAt: '2021-07-10T00:44:09.386Z',
    updatedAt: '2021-07-10T00:44:09.386Z',
  },
  {
    id: '3e7ca498-8646-4493-bce6-b8cad8c5908e',
    address: 'Address 2',
    name: 'Store name 2',
    description: 'Description 2',
    ownerId: '7d13c742-cd2d-45b7-9866-20dbb0eaa2e9',
    picture: 'https://cdn.geekwire.com/wp-content/uploads/2020/02/amzngo1.jpeg',
    createdAt: '2021-07-10T00:44:09.386Z',
    updatedAt: '2021-07-10T00:44:09.386Z',
  },
  {
    id: '9440a98a-2973-484b-b8bd-78beb6b24eef',
    address: 'Address 3',
    name: 'Store name 3',
    description: 'Description 3',
    ownerId: '26cbed2e-b38c-461e-a387-977a6cbff6dc',
    picture: null,
    createdAt: '2021-07-10T00:44:09.386Z',
    updatedAt: '2021-07-10T00:44:09.386Z',
  },
];

describe('StoresList', () => {
  describe('When user is not logged-in', () => {
    it('renders the stores list', async () => {
      let storesRenderer;
      apiGet.mockImplementation(() => ({
        data: storesResponse,
        statusCode: 200,
      }));
      await renderer.act(async () => {
        storesRenderer = await renderer.create(<TestRouter path="/stores" />);
      });
      const tree = storesRenderer.toJSON();
      expect(tree).toMatchSnapshot();
    });

    afterEach(() => {
      cleanup();
      jest.resetAllMocks();
    });
  });
});
