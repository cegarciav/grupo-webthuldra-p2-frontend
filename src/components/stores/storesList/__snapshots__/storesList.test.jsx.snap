// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoresList When user is not logged-in renders the stores list 1`] = `
<main>
  <h1>
    Lista de tiendas
  </h1>
  <section
    className="searchbar-container"
  >
    <input
      className="main-searchbar"
      onChange={[Function]}
      placeholder="Buscar tiendas"
      type="text"
      value=""
    />
  </section>
  <ul
    className="stores-list"
  >
    <li>
      <a
        className="store-summary"
        href="/stores/6f4e5985-2ff8-4118-bbaf-e6004d757c13"
        onClick={[Function]}
      >
        <article>
          <h1>
            Store name 1
          </h1>
          <section>
            <img
              alt="Description 1"
              src="/default_store_picture.png"
            />
          </section>
          <p>
            Dirección: Address 1
          </p>
        </article>
      </a>
    </li>
    <li>
      <a
        className="store-summary"
        href="/stores/3e7ca498-8646-4493-bce6-b8cad8c5908e"
        onClick={[Function]}
      >
        <article>
          <h1>
            Store name 2
          </h1>
          <section>
            <img
              alt="Description 2"
              src="https://cdn.geekwire.com/wp-content/uploads/2020/02/amzngo1.jpeg"
            />
          </section>
          <p>
            Dirección: Address 2
          </p>
        </article>
      </a>
    </li>
    <li>
      <a
        className="store-summary"
        href="/stores/9440a98a-2973-484b-b8bd-78beb6b24eef"
        onClick={[Function]}
      >
        <article>
          <h1>
            Store name 3
          </h1>
          <section>
            <img
              alt="Description 3"
              src="/default_store_picture.png"
            />
          </section>
          <p>
            Dirección: Address 3
          </p>
        </article>
      </a>
    </li>
  </ul>
</main>
`;
